### DISCUSSION TITLE
what's the right valid access token and appropriate scopes for publishing one GitHub package?

### DISCUSSION THREAD

**Discussion by xgqfrms** [community/community#28844] [Actions] [answered]
# what's the right valid access token and appropriate scopes for publishing one GitHub package?

> `.npmrc`

```diff
- //npm.pkg.github.com/:_authToken=GITHUB_TOKEN
+ //npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
@xgqfrms:registry=https://npm.pkg.github.com
```

## what's wrong with that? I think the scope is appropriate for publish the package.

![image](https://user-images.githubusercontent.com/7291672/183575590-3d707cae-d9ad-422a-aec7-8d2b6ada95de.png)

## error message

https://github.com/xgqfrms/wcui/runs/7739940709?check_suite_focus=true#step:5:28

![image](https://user-images.githubusercontent.com/7291672/183576104-3e89cd75-ea57-4520-9c4a-f81b150de232.png)

**Comment by xgqfrms** [comment]
## Q1: what's the right way of `GITHUB_TOKEN` usage? 



1. without `${}` symbol wrapped

```code
//npm.pkg.github.com/:_authToken=GITHUB_TOKEN
@xgqfrms:registry=https://npm.pkg.github.com

```

2. with `${}` symbol wrapped


```code
//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
@xgqfrms:registry=https://npm.pkg.github.com

```


## Q2: where do the secrets `NPM_TOKEN` come from for GitHub Action workflow? 


```yml
   NODE_AUTH_TOKEN: ${{secrets.NPM_TOKEN}}

```


1.  copy from `npm tokens`

https://www.npmjs.com/settings/xgqfrms/tokens


3. just same as the `GITHUB_TOKEN`, which is GitHub Access Token

https://github.com/xgqfrms/wcui/blob/main/.github/workflows/npm-publish-github-packages.yml#L36

**Comment by airtower-luna** [comment]
> Q1: what's the right way of GITHUB_TOKEN usage?

Definitely with `${}`, that's the syntax for accessing environment variables (like the `GITHUB_TOKEN`).

> Q2: where do the secrets NPM_TOKEN come from for GitHub Action workflow?

Considering you're pushing to `npm.pkg.github.com` you need a token from GitHub. According to [the documentation](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry) the `GITHUB_TOKEN` should work, but you'll need a PAT if you also want to download other packages from there.

**Reply by xgqfrms** [reply]
PAT === Personal access token

**Comment by xgqfrms** [comment]
# can not create a token named `GITHUB_TOKEN`

<img width="664" alt="image" src="https://user-images.githubusercontent.com/7291672/183590123-9cd6fa37-71c8-43dc-a387-49d4c75fb03b.png">

## Q3: Is `GITHUB_TOKEN` the default keyword or do I need to create a `PAT` named `GITHUB_TOKEN`?

## A3:

<img width="781" alt="image" src="https://user-images.githubusercontent.com/7291672/183614585-a32bba86-33b8-4148-8c3d-25e0889b2626.png">


https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret

https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#setting-the-permissions-of-the-github_token-for-your-repository

https://docs.github.com/en/rest/overview/permissions-required-for-github-apps


# GitHub App

https://github.com/settings/apps/new

<img width="823" alt="image" src="https://user-images.githubusercontent.com/7291672/183616472-972ca648-0941-41fe-93c0-a15057226c95.png">

**Comment by xgqfrms** [comment]
## Everything looks fine, but why still can't publish the GitHub package?

```code
//npm.pkg.github.com/:_authToken=${NPM_GITHUB_TOKEN}
@xgqfrms:registry=https://npm.pkg.github.com/
```

https://github.com/xgqfrms/wcui/blob/main/.npmrc


```yml
        env:
          # NPM_TOKENN for publish npm package ✅ https://registry.npmjs.org/
          # NODE_AUTH_TOKEN: ${{secrets.NPM_TOKEN}}
          # GITHUB_TOKEN for publish  GitHub Package ✅ registry=https://npm.pkg.github.com
          NODE_AUTH_TOKEN: ${{secrets.NPM_GITHUB_TOKEN}}
```



https://github.com/xgqfrms/wcui/blob/main/.github/workflows/npm-publish-github-packages.yml

**Comment by xgqfrms** [comment]
# solution ?


1. first create one `PAT`

https://github.com/settings/tokens


<img width="850" alt="image" src="https://user-images.githubusercontent.com/7291672/183594402-ebe4256f-d5ec-4e4b-ac7c-d5ea8a6cba99.png">


3. Use the local Terminal login with `PAT` as the password

```js
$  npm login --registry=https://npm.pkg.github.com
npm notice Log in on https://npm.pkg.github.com/
Username: xgqfrms
Password: 
Email: (this IS public) xgqfrms@xgqfrms.xyz
Logged in as xgqfrms on https://npm.pkg.github.com/.

```

<img width="833" alt="image" src="https://user-images.githubusercontent.com/7291672/183593761-0f352b6a-5c69-4e1a-9b8d-10ca00d1bca1.png">

??? scoped

```sh
$ npm login --registry=https://npm.pkg.github.com --scope=@xgqfrms

```

<img width="656" alt="image" src="https://user-images.githubusercontent.com/7291672/183597017-46c6cb6a-48b6-47b2-92f0-22538b324f30.png">



3. try create a new release ❌

```md
🚀 publish my first github package ✅ with real PAT token

```

4. local publish ❌ 

<img width="904" alt="image" src="https://user-images.githubusercontent.com/7291672/183596276-3fb10dd1-122b-4129-8807-aa71ede94299.png">

**Comment by xgqfrms** [comment]
# ???

```yml
- //npm.pkg.github.com/:_authToken=GITHUB_TOKEN
- //npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
+ //npm.pkg.github.com/:_authToken={GITHUB_TOKEN}
@xgqfrms:registry=https://npm.pkg.github.com
```

https://github.com/xgqfrms/wcui/blob/main/.npmrc

https://github.com/pnpm/pnpm/issues/2687

**Comment by xgqfrms** [comment]
# GitHub Packages `only` supports `scoped` npm packages.

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry#publishing-a-package

<img width="800" alt="image" src="https://user-images.githubusercontent.com/7291672/183603910-6369b178-7b2b-4fdb-b6af-579f4b515160.png">

## `GITHUB_TOKEN`

> GitHub Packages  use `GITHUB_TOKEN` error


https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow

**Comment by xgqfrms** [comment, marked as answer]
# solution ✅

1. login with scoped

```sh
$ npm login --registry=https://npm.pkg.github.com --scope=@xgqfrms

```

2.  add the pure string `PAT` as token 🚀

add `.npmrc` to git ignore for security

```code
//npm.pkg.github.com/:_authToken=JUST_PUT_YOUR_PAT_STRING_IN_HERE
@xgqfrms:registry=https://npm.pkg.github.com/

```

3. publish in the local terminal

<img width="536" alt="image" src="https://user-images.githubusercontent.com/7291672/183605198-a14e6cc5-be7d-4e11-9766-bfc1239e2a48.png">


## refs

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry#authenticating-with-a-personal-access-token

<img width="820" alt="image" src="https://user-images.githubusercontent.com/7291672/183618163-ea49dbee-663a-4bea-a95b-0ae8e24971c7.png">
